pygem/setup/config.py

```diff
@@ -7,42 +7,336 @@
+        self.ensure_config()
     
-    if (not isfile) or (overwrite):
-        os.makedirs(basedir, exist_ok=True)  # Ensure the base directory exists
-        shutil.copy(source_config_file, config_file)  # Copy the file
-        print(f"Copied default configuration to {config_file}")
+    def ensure_config(self):
+        """Ensure the configuration file exists, creating or overwriting it if necessary"""
+        if not os.path.isfile(self.config_path) or self.overwrite:
+            self.copy_source_config()
+
+    def copy_source_config(self):
+        """Copy the default configuration file to the expected location"""
+
+        os.makedirs(self.base_dir, exist_ok=True)
+        shutil.copy(self.source_config_path, self.config_path)
+        print(f"Copied default configuration to {self.config_path}")
+        
+    def read_config(self, validate=True):
+        """Read the configuration file and return its contents as a dictionary while preserving formatting."""
+        ryaml = ruamel.yaml.YAML()
+        with open(self.config_path, 'r') as f:
+            user_config = ryaml.load(f)  # Using ruamel.yaml for preservation
+
+        if validate:
+            self.validate_config(user_config)
+
+        return user_config
+
+    def write_config(self, config):
+        """Write the configuration dictionary to the file while preserving quotes."""
+        ryaml = ruamel.yaml.YAML()
+        ryaml.preserve_quotes = True
+        with open(self.config_path, 'w') as file:
+            ryaml.dump(config, file)  # This will preserve quotes
+    
+    def update_config(self, updates):
+        """Update multiple keys in the YAML configuration file while preserving quotes and original types."""
+        config = self.read_config(validate=False)  # Read existing config
+        
+        for key, value in updates.items():
+            if key not in self.EXPECTED_TYPES:
+                raise KeyError(f"Unrecognized configuration key: {key}")
+            keys = key.split('.')
+            d = config
+            for sub_key in keys[:-1]:
+                d = d[sub_key]
+
+            d[keys[-1]] = value
+        
+        self.validate_config(config)
+        self.write_config(config)
+    
+    def validate_config(self, data):
+        """Validate the configuration file against expected types and required keys"""
+        for key, expected_type in self.EXPECTED_TYPES.items():
+            keys = key.split(".")
+            sub_data = data
+            for sub_key in keys:
+                if isinstance(sub_data, dict) and sub_key in sub_data:
+                    sub_data = sub_data[sub_key]
+                else:
+                    raise KeyError(f"Missing required key in configuration: {key}")
+
+            if not isinstance(sub_data, expected_type):
+                raise TypeError(f"Invalid type for '{key}': expected {expected_type}, not {type(sub_data)}")
+
+            # Check elements inside lists (if defined)
```

Comment:
I'd make this more descriptive, something like
```suggestion
            # For lists, check that all elements of the list are of expected type
```